### PR TITLE
DomainKeys Identified Mail (DKIM) Signatures

### PR DESCRIPTION
**Pull request by yaddran**
Implements #23493

This is a simple feature that exists in PHPMailer for years now. It helps a lot in preventing mails sent being classified as SPAM.

### CODE REVIEW

**Comment by yaddran** [comment]
Done:
- opened a feature request number #23493
- referenced the PR from the issues and vice versa 
- added email_dkim prefix
- added the possibility to specify key as value

This feature is completely connected to email so however you treat smtp settings, for instance, should be the same way how you treat the DKIM settings. Since the email settings are using example.com as defaults and DKIM signature has to use the same values I think it is wise to have the same values example.com values for DKIM.

The validation is not necessary. The email sending will not stop. If one of the needed values are skipped or wrong, the PHPMailer will still send mail (silently skipping the problem) but it could happen that the signature is not added or that the signature later does not pass the DKIM check at the arriving mail server. I think that this is acceptable behavior. When you are trying to configure DKIM it is logical that you will check if it is working.

Todo:
- update documentation (a small pointer please :) )

File: core/email_api.php [hunk 1]
```diff
@@ -1300,6 +1300,16 @@ function email_send( EmailData $p_email_data ) {
 			break;
 	}
 
+	#apply DKIM settings
+	if( config_get( 'dkim_enable' ) ) {
+		$t_mail->DKIM_domain = config_get( 'email_dkim_domain' );
+		$t_mail->DKIM_private = config_get( 'email_dkim_private_key' );
```
Comment (atrol) [inline core/email_api.php]: `private_key` is certainly better than `private`, but wouldn't something like `private_key_file_path` be even better?

**Comment by atrol** [comment]
> update documentation (a small pointer please :) )

Options should be added to https://github.com/mantisbt/mantisbt/blob/master/docbook/Admin_Guide/en-US/config/email.xml

Not complete up to date, but should help
https://www.mantisbt.org/wiki/doku.php/mantisbt:docbook

File: core/email_api.php [hunk 2]
```diff
@@ -1303,7 +1303,7 @@ function email_send( EmailData $p_email_data ) {
 	#apply DKIM settings
 	if( config_get( 'dkim_enable' ) ) {
 		$t_mail->DKIM_domain = config_get( 'email_dkim_domain' );
-		$t_mail->DKIM_private = config_get( 'email_dkim_private_key' );
+		$t_mail->DKIM_private = config_get( '$g_email_dkim_private_key_file_path' );
```
Comment (atrol) [inline core/email_api.php]: remove the `$g_`

**Comment by atrol** [comment]
You added the documentation after `$g_smtp_port` but before `$g_smtp_connection_mode`, thus splitting the smtp settings.
The settings should be added after `$g_smtp_password`.

After applying this change and squashing the commits, the PR is fine from my side.
@vboctor might want some more changes.
Thanks for your contribution.

File: config_defaults_inc.php [hunk 3]
```diff
@@ -680,6 +680,82 @@
...
+ * @global string $g_email_dkim_domain
+ */
+$g_email_dkim_domain = 'example.com';
+
+/**
+ * DomainKeys Identified Mail (DKIM) Signatures private key path
+ * Path to the private key. If $g_email_dkim_private_key_string is specified
+ * this setting will not be used.
+ * @see $g_email_dkim_private_key_string
+ * @see $g_dkim_enable
```
Comment (vboctor) [inline config_defaults_inc.php]: Wrong config option name.

File: config_defaults_inc.php [hunk 4]
```diff
@@ -680,6 +680,82 @@
...
+ */
+$g_email_dkim_private_key_file_path = '/path/to/readable/example.com/private_key.pem';
+
+
+/**
+ * DomainKeys Identified Mail (DKIM) Signatures private key value
+ * This string should contain private key for signing. Leave empty
+ * string if you wish to load the key from the file defined with
+ * $g_email_dkim_private_key_file_path.
+ * @see $g_dkim_enable
```
Comment (vboctor) [inline config_defaults_inc.php]: Wrong config option name.

File: core/email_api.php [hunk 5]
```diff
@@ -1300,6 +1300,16 @@ function email_send( EmailData $p_email_data ) {
 			break;
 	}
 
+	#apply DKIM settings
+	if( config_get( 'dkim_enable' ) ) {
+		$t_mail->DKIM_domain = config_get( 'email_dkim_domain' );
+		$t_mail->DKIM_private = config_get( 'email_dkim_private_key_file_path' );
```
Comment (vboctor) [inline core/email_api.php]: I suggest only setting the file path if the key string is not set.

File: config_defaults_inc.php [hunk 6]
```diff
@@ -680,6 +680,82 @@
...
+
+/**
+ * DomainKeys Identified Mail (DKIM) Signatures private key path
+ * Path to the private key. If $g_email_dkim_private_key_string is specified
+ * this setting will not be used.
+ * @see $g_email_dkim_private_key_string
+ * @see $g_dkim_enable
+ * @global string $g_email_dkim_private_key_file_path
+ */
+$g_email_dkim_private_key_file_path = '/path/to/readable/example.com/private_key.pem';
```
Comment (vboctor) [inline config_defaults_inc.php]: Specially now that we have a file path or a key string, let's default the file path to an empty string.

File: config_defaults_inc.php (hunk 4 above)
Comment (atrol) [reply config_defaults_inc.php]: It's a `@see` tag

File: core/email_api.php (hunk 5 above)
Comment (yaddran) [reply core/email_api.php]: Why adding extra code when this is already handled by PHPMailer. We have specified in the documentation and comments the order the settings are applied in.

**Comment by yaddran** [comment]
Hm... trouble with squash.... :(

**Comment by atrol** [comment]
> Hm... trouble with squash.... :(
> How do I actually do this... ?

```
Use git rebase -i <after-this-commit> and replace "pick" on the second and subsequent commits with "squash" or "fixup
```
https://stackoverflow.com/questions/5189560/squash-my-last-x-commits-together-using-git

**Comment by yaddran** [comment]
hm... exactly what I did but then when I tried to push it complained that I need to pull first and them merge came up and then it came up as a new commit... :(
I should probably do git push --force after ?

**Comment by atrol** [comment]
> I should probably do git push --force after ?

Right

**Comment by atrol** [comment]
Closed in favor of https://github.com/mantisbt/mantisbt/pull/1213

**Pull request closed without merging**